go/pkg/casng/batching.go

```diff
@@ -314,12 +318,113 @@ func (u *uploader) writeBytes(ctx context.Context, name string, r io.Reader, siz
 //
 // This method must not be called after cancelling the uploader's context.
 func (u *BatchingUploader) Upload(ctx context.Context, reqs ...UploadRequest) ([]digest.Digest, Stats, error) {
-	panic("not yet implemented")
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload; reqs=%d", len(reqs))
+	defer contextmd.Infof(ctx, log.Level(1), "[casng] upload.done; reqs=%d", len(reqs))
+
+	var stats Stats
+
+	if len(reqs) == 0 {
+		return nil, stats, nil
+	}
+
+	var undigested []UploadRequest
+	digested := make(map[digest.Digest]UploadRequest)
+	var digests []digest.Digest
+	for _, r := range reqs {
+		if r.Digest.Hash == "" {
+			undigested = append(undigested, r)
+			continue
+		}
+		digested[r.Digest] = r
+		digests = append(digests, r.Digest)
+	}
+	missing, err := u.MissingBlobs(ctx, digests)
+	if err != nil {
+		return nil, stats, err
+	}
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload; missing=%d, undigested=%d", len(missing), len(undigested))
+
+	reqs = undigested
+	for _, d := range missing {
+		reqs = append(reqs, digested[d])
+		delete(digested, d)
+	}
+	for d := range digested {
+		stats.BytesRequested += d.Size
+		stats.LogicalBytesCached += d.Size
+		stats.CacheHitCount += 1
+		stats.DigestCount += 1
+	}
+	if len(reqs) == 0 {
+		contextmd.Infof(ctx, log.Level(1), "[casng] upload: nothing is missing")
+		return nil, stats, nil
+	}
+
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload: uploading; reqs=%d", len(reqs))
+	ch := make(chan UploadRequest)
+	resCh := u.streamPipe(ctx, ch)
+
+	u.clientSenderWg.Add(1)
+	go func() {
+		defer close(ch) // let the streamer terminate.
+		defer u.clientSenderWg.Done()
+
+		contextmd.Infof(ctx, log.Level(1), "[casng] upload.sender.start")
+		defer contextmd.Infof(ctx, log.Level(1), "[casng] upload.sender.stop")
+
+		for _, r := range reqs {
+			select {
+			case ch <- r:
+			case <-u.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	var uploaded []digest.Digest
+	for r := range resCh {
```

Comment:
How is `resCh` closed?

Reply:
`streamPipe` takes an input channel `ch` and returns an output channel `resCh`. The caller (here) owns the input channel while `streamPipe` owns the output channel. When `ch` is closed (line 369), `streamPipe` will close `resCh` after it's done processing messages received on `ch`.